server/src/main/scala/org/http4s/server/middleware/authentication/NonceKeeperF.scala

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2014 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.server.middleware.authentication
+
+import cats.effect.Clock
+import cats.effect.Concurrent
+import cats.effect.Timer
+import cats.effect.concurrent.Ref
+import cats.effect.concurrent.Semaphore
+import cats.syntax.all._
+
+import java.util.LinkedHashMap
+import java.{util => ju}
+import scala.concurrent.duration.MILLISECONDS
+
+private[authentication] object NonceKeeperF {
+  def apply[F[_]: Timer](
+      staleTimeout: Long,
+      nonceCleanupInterval: Long,
+      bits: Int,
+  )(implicit F: Concurrent[F]): F[NonceKeeperF[F]] = for {
+    semaphore <- Semaphore[F](1)
+    currentTime <- Clock[F].realTime(MILLISECONDS)
+    lastCleanup <- Ref[F].of(currentTime)
+  } yield new NonceKeeperF(staleTimeout, nonceCleanupInterval, bits, semaphore, lastCleanup)
+}
+
+/** A thread-safe class used to manage a database of nonces.
+  *
+  * @param staleTimeout Amount of time (in milliseconds) after which a nonce
+  *                     is considered stale (i.e. not used for authentication
+  *                     purposes anymore).
+  * @param bits The number of random bits a nonce should consist of.
+  */
+private[authentication] class NonceKeeperF[F[_]: Timer](
+    staleTimeout: Long,
+    nonceCleanupInterval: Long,
+    bits: Int,
+    semaphore: Semaphore[F],
+    lastCleanup: Ref[F, Long],
+)(implicit F: Concurrent[F]) {
+  require(bits > 0, "Please supply a positive integer for bits.")
+  private val nonces = new LinkedHashMap[String, NonceF[F]]
+
+  val clock = Clock[F]
+
+  /** Removes nonces that are older than staleTimeout
+    * Note: this _MUST_ be executed with the singleton permit from the semaphore
+    */
+  private def checkStale(): F[Unit] =
```

Comment:
Scaladoc is good but this really gets it across!

```suggestion
  private def unsafeCheckStale(): F[Unit] =
```

Reply:
I've always had `unsafe` mapped to "side-effecting without F[_]", but I guess the only way to make this safe is to enrich the semaphore with a value, like `Resource`. Good suggestion, that comment was bugging me as well.